src/Compilers/Core/Portable/CommandLine/ErrorLogger.cs

```diff
@@ -211,23 +280,108 @@ private static string GetKind(DiagnosticSeverity severity)
                     return "error";
 
                 case DiagnosticSeverity.Warning:
+                    return "warning";
+
                 case DiagnosticSeverity.Hidden:
                 default:
-                    // note that in the hidden or default cases, we still write out the actual severity as a
-                    // property so no information is lost. We have to conform to the SARIF spec for kind,
-                    // which allows only pass, warning, error, or notApplicable.
-                    return "warning";
+                    // hidden diagnostics are not reported on the command line and therefore not currently given to 
+                    // the error logger. We could represent it with a custom property in the SARIF log if that changes.
+                    Debug.Assert(false);
+                    goto case DiagnosticSeverity.Warning;
             }
         }
 
         public void Dispose()
         {
             _writer.WriteArrayEnd();  // results
-            _writer.WriteObjectEnd(); // runLog
-            _writer.WriteArrayEnd();  // runLogs
+
+            WriteRules();
+
+            _writer.WriteObjectEnd(); // run
+            _writer.WriteArrayEnd();  // runs
             _writer.WriteObjectEnd(); // root
             _writer.Dispose();
         }
+
+        /// <summary>
+        /// Represents a distinct set of <see cref="DiagnosticDescriptor"/>s and provides unique string keys 
+        /// to distinguish them.
+        ///
+        /// The first <see cref="DiagnosticDescriptor"/> added with a given <see cref="DiagnosticDescriptor.Id"/>
+        /// value is given that value as its unique key. Subsequent adds with the same ID will have .NNN
+        /// apppended to their with an auto-incremented numeric value.
+        /// </summary>
+        private sealed class DiagnosticDescriptorSet
+        {
+            // DiagnosticDescriptor.Id -> auto-incremented counter
+            private Dictionary<string, int> _counters = new Dictionary<string, int>();
+
+            // DiagnosticDescriptor -> unique key
+            private Dictionary<DiagnosticDescriptor, string> _keys = new Dictionary<DiagnosticDescriptor, string>();
```

Comment:
DiagnosticDescriptor is IEquatable<DiagnosticDescriptor> by value, so that's why I thought default comparer would work. However, I see now that common case of unequal descriptors with same ID is when they differ only by message format and we don't write out the message format. So I do need a comparer that only compares the parts we write out. On it.